compiler/generator_java.c

```diff
@@ -272,7 +272,7 @@ static void generate_AE(struct generator * g, struct node * p) {
             break;
         case c_len: /* Same as size() for Java. */
         case c_size:
-            w(g, "current.length()");
+            w(g, "limit");
```

Comment:
There look to be a few other places where `limit` is assumed to be the full currently length of the string.  It can definitely be the wrong value here, but it's possible for some that `limit` will always get restored to be the same value (or at least should be - I tried to test this situation but actually managed to push limit **above** the original length of the string which I think must be a bug so I'll need to investigate more).  Either way I think we need to track the current length separately to `limit` so we might as well use it in the other places as it's more conceptually correct.

Reply:
I fixed the length to be separate from limit, existing tests pass and also your demonstration above now yields `test`